src/ogr/feature.jl

```diff
@@ -403,12 +447,14 @@ const _FETCHFIELD = Dict{OGRFieldType,Function}(
 )
 
 function getfield(feature::Feature, i::Integer)
-    return if isfieldset(feature, i)
-        _fieldtype = gettype(getfielddefn(feature, i))
+    return if !isfieldset(feature, i)
+        getdefault(feature, i)
```

Comment:
This still seems wrong to me. If the field is not set it is `nothing`. It should not be displayed at all. 

See also https://gdal.org/development/rfc/rfc67_nullfieldvalues.html

However, I can imagine we make it `missing` as well, which would be a bit more user friendly and it is one of the options presented in the above RFC.

Reply:
That's great feedback, thanks!

Let me go through the following RFCs and functions (and any other corresponding associated functions):
* https://gdal.org/development/rfc/rfc53_ogr_notnull_default.html
  * `getdefault(...)`
  * `setdefault!(...)`
  * `setnullable!(...)`
  * `isdefaultdriverspecific(...)`
  * `validate()`
* https://gdal.org/development/rfc/rfc67_nullfieldvalues.html
  * `getfield(feature, i)`
  * `isfieldnull(feature, i)`
  * `isfieldsetandnotnull(feature, i)`
  * `setfieldnull!(feature, i)`

I will
* provide a corresponding discussion under https://yeesian.com/ArchGDAL.jl/stable/considerations/ as part of this PR,
* update the PR title and so on to reflect the design decision, before
* updating the implementation and tests to reflect it.

Reply:
Here's what I have as an early draft:

```
+-------+---------+---------+
|       |  null   | notnull |
+-------+---------+---------+
| set   | missing | value   |
+-------+---------+---------+
| unset | nothing | nothing |
+-------+---------+---------+
```

The design decision reflects that
* a value can either "not exist" or "be unknown", but not both.
* "set" means that a value "exists" (even if it might be unknown): the corresponding value we use for it is `missing`.
* "null" means that a value is "unknown" (presuming that it exists): the corresponding value we use for it is `nothing`.

Reply:
In RFC 67, under [Compatibility](https://gdal.org/development/rfc/rfc67_nullfieldvalues.html#compatibility-issues), it says

>All code, in GDAL source code, and in calling external code, that currently uses OGRFeature::IsFieldSet() / OGR_F_IsFieldSet() should also be updated to used IsFieldNull() / OGR_F_IsFieldNull(), **either to act exactly as in the unset case**, or add a new appropriate behavior.

Which would mean that we could use `missing` also in the unset case as well. 

While I think `nothing` is semantically more correct, I fear that having a `Union{Nothing, Missing, ValueType}` is not pragmatic. We should test this with a few special files with both unset as null fields.

Reply:
> We should test this with a few special files

Let me know If there are any cases you wish to call out or files you'd like to ensure gets tested.

Reply:
I would say from my reading of the [RFC 67](https://github.com/OSGeo/gdal/blob/master/gdal/doc/source/development/rfc/rfc67_nullfieldvalues.rst) and the GDAL source files containing nMarker1, that the logic is the following: `OGRUnsetMarker` and `OGRNullMarker`are mutually exclusive.

I propose the following mapping:

**nMarker1/2/3**|**Field has default value ?**|**Returned value**
:-----:|:-----:|:-----:
OGRUnsetMarker|true|default value
OGRUnsetMarker|false|missing
OGRNullMarker|true / false|nothing
other value|true / false|value

Reply:
Thanks Mathieu! What do you think of the following proposal?

```
+-------+---------+--------------+
|       |  null   |   notnull    |
+-------+---------+--------------+
| set   | missing |    value     |
+-------+---------+--------------+
| unset | nothing | getdefault() |
+-------+---------+--------------+
```
where `getdefault()` should return `missing` if there is no default value.

I think that way,
* "NULL" has flavours of `missing` and `nothing` depending on whether the field is set.
* "NOTNULL" is consistent with "Note that adding a field with a NOT NULL constraint on a non-empty layer is generally impossible, unless a DEFAULT value is associated with it" and "Fields with NOT NULL constraints are sometimes accompanied with a DEFAULT clause so as to be able to create a new feature without filling all fields, while maintaining integrity" in https://gdal.org/development/rfc/rfc53_ogr_notnull_default.html.
* Fields that are set can propagate `missing` value behavior (rather than throwing an error on `nothing`). Users can use [`isfieldset()` and `unsetfield!(...)`](https://github.com/yeesian/ArchGDAL.jl/blob/d177575acd70775434f4ace21b9313573823469c/src/ogr/feature.jl#L126-L150) to check for this behavior and control it to return `nothing` (rather than `missing`) if they want.
* Fields that might be unset have a `FillUnsetWithDefault()` method in GDAL (that should be rarely used; per https://gdal.org/development/rfc/rfc53_ogr_notnull_default.html).

> I would say from my reading of the RFC 67 and the GDAL source files containing nMarker1, that the logic is the following: OGRUnsetMarker and OGRNullMarkerare mutually exclusive.

It's not obvious to me that `OGRUnsetMarker ` and `OGRNullMarkerare` are mutually exclusive -- can you explain how you came to the conclusion?

Reply:
@evetion I'm afraid the proposal is a complex one, but it is not trying to be more complex than it needs to be. From what I've read of the GDAL RFCs, I think that `missing`, `nothing` or `getdefault()` can be sensible behavior depending on the circumstance.

Reply:
From my reading of the RFCs, `getdefault()` is only used for **insertion**, when a field can't be null, but isn't given. This code is about reading, so it shouldn't come into play at all.

So your original proposal seems right:
| read field | null    | notnull |
|------------|---------|---------|
| set        | missing | value   |
| unset      | nothing | nothing |

Reply:
@yeesian 

> > I would say from my reading of the RFC 67 and the GDAL source files containing nMarker1, that the logic is the following: OGRUnsetMarker and OGRNullMarke rare mutually exclusive.
> 
> It's not obvious to me that OGRUnsetMarker and OGRNullMarkerare are mutually exclusive -- can you explain how you came to the conclusion?

Well, my understanding of OGR cpp code may be superficial (1st time I looked into it), but here it is for comments:

1. An `ÒGRField` is defined by an [Union](https://github.com/OSGeo/gdal/blob/814434983f73c2917ceb36ee8a72fa3225ce646c/gdal/ogr/ogr_core.h#L730-L788)
2. In this Union, there is as [`Set`](https://github.com/OSGeo/gdal/blob/814434983f73c2917ceb36ee8a72fa3225ce646c/gdal/ogr/ogr_core.h#L770-L774) struct containing the three markers `nMarker1`, `nMarker2`, `nMarker3`
3. If one of the members (other than the `Set` struc member) of an `OGRField` is set, `nmarker2` and `nmarker3` are set to `0`
When `nMarker1` is set, `nMarker2` and `nMarker3` are set equal to `nMarker1`. 
`nMarker1` is set to [two possible values](https://github.com/OSGeo/gdal/blob/814434983f73c2917ceb36ee8a72fa3225ce646c/gdal/ogr/ogr_core.h#L715-L728) by two functions:
    - `OGRUnsetMarker` by [`OGR_RawField_SetUnset`](https://github.com/OSGeo/gdal/blob/b4c3e127edda6cbb0c8e3f37d287e624c7d84395/gdal/ogr/ogrfeature.cpp#L6968-L6973)
    - `OGRNullMarker`by [`OGR_RawField_SetNull`](https://github.com/OSGeo/gdal/blob/b4c3e127edda6cbb0c8e3f37d287e624c7d84395/gdal/ogr/ogrfeature.cpp#L6992-L6997)

Therefore since Unset state and Null state for an OGRField are implemented by two different values of one unique variable, the two states are mutually exclusive

Reply:
@mathieu17g Did you switch the meanings of `missing` and `nothing` by accident? IMHO they should be switched. The Julia manual states: `missing is equivalent to NULL in SQL`, while `[nothing is] used by convention when there is no value to return`.

Then you'd get.

Read Field | null | notnull
-- | -- | --
set | missing | value
unset | n/a | nothing


Write Field | nullable | notnullable
-- | -- | --
nothing | unset | unset
missing | null | default value
value | value | value

Reply:
@evetion: Well my proposal was intended but mostly based on intuition.
Following your remark, I have dug a bit more in Julia's documentation about missing and nothing values and SQL null usage, I have found some arguments to support my convention proposition:
- `OGRField.Set.nMarker1 = OGRUnsetMarker`  to Julia `missing` and 
- `OGRField.Set.nMarker1 = OGRNullMarker` to Julia `nothing`

But none of these arguments is definitive and one may find some other arguments supporting the opposite convention. So since, we have to decide to be able to move on, **both conventions proposals are fine for me**. And if later, we encounter a similar convention issue in another Julia package with convincing argument, we will be able to easily change the convention in ArchGDAL.

---

If you have spare time, here are below, a few additional observations based on vague pieces of docs found on the topic in Julia's documentation

### In Julia's [documentation page on missing values](https://docs.julialang.org/en/v1/manual/missing/#missing):

> Julia provides support for representing missing values in the statistical sense, that is for situations where no value is available for a variable in an observation, but a valid value theoretically exists. Missing values are represented via the missing object, which is the singleton instance of the type Missing. **missing is equivalent to NULL in SQL and NA in R**, and behaves like them in most situations.

One interpretation could be:
- NULL in SQL is associated with no value provided for a field when the record was written. So when one does not set a value in a record, it is always set to NULL
- The equivalent in OGR is unset since a feature fields are initiated with the OGRField set to OGRUnsetMarker in the [basic constructor](https://github.com/OSGeo/gdal/blob/b4c3e127edda6cbb0c8e3f37d287e624c7d84395/gdal/ogr/ogrfeature.cpp#L99-L107).
- The `OGRNullMarker` is never set by default in OGR, it will be set on purpose with the function `SetFieldNull` on a particular OGRField instance

**So my understanding of this part of Julia documentation is that it could also be in favor of associating `OGRField.Set.nMarker1 = OGRUnsetMarker`  to Julia `missing`**

### In Julia's documentation page on Function and in the paragraph ["Returning nothing"](https://docs.julialang.org/en/v1/manual/functions/#Returning-nothing)

> For **functions that do not need to return a value** (functions used only for some side effects), the Julia convention is to return the value nothing

Here with `ArchGDAL.getfield` the main purpose is not any side effect, but to return a supposedly existing value of the field

I have not found your reference:

> [nothing is] used by convention when there is no value to return

### In Julia's documents FAQ's paragraph ["Nothingness and missing values"](https://docs.julialang.org/en/v1/manual/faq/#Nothingness-and-missing-values)

> For situations where a value `x` of type `T` exists only sometimes, the `Union{T, Nothing}` type can be used for function arguments, object fields and array element types as the equivalent of [Nullable](https://en.wikipedia.org/wiki/Nullable_type), [Option](https://en.wikipedia.org/wiki/Nullable_type) or [Maybe](https://en.wikipedia.org/wiki/Nullable_type) in other languages. 

When I follow the link to the wikipedia page, the explinations given there are confusing since they never refer to existence of something similar to Julia's dual "missingness" and "nothingness"
The rest of the paragraph does not allow to  decide.

Reply:
It seems like we're getting close to consensus here. @evetion, will you be okay if I update my original proposal in https://github.com/yeesian/ArchGDAL.jl/pull/238#discussion_r706615531 and yours in https://github.com/yeesian/ArchGDAL.jl/pull/238#discussion_r707369926 to the following (changes are marked with an asterisk):

```
+-------+---------+----------+
| Read  |  null   | notnull  |
+-------+---------+----------+
| set   | missing |  value   |
+-------+---------+----------+
| unset | nothing | missing* |
+-------+---------+----------+

+---------+-------------+--------------+
|  Write  |  nullable   | notnullable  |
+---------+-------------+--------------+
| nothing |    unset    |     unset    |
+---------+-------------+--------------+
| missing |     null    | getdefault() |
+---------+-------------+--------------+
|  value  |    value    |    value     |
+---------+-------------+--------------+

```

Namely, we
1. There is no change to your proposal when writing fields.
2. When reading fields, we return `missing` instead of `nothing` for fields that are `(unset, notnull)` case.

I'll start reflecting this discussion and decision in the documents, so there's still time to make changes to intended behavior.

Reply:
@yeesian I'm fine with your proposed change, it seems more pragmatic to me. @mathieu17g Thanks for all the thought you put into the discussion, it has made things much better and more explicit.

Reply:
@mathieu17g you have saved me from potential hours of head-scratching with https://github.com/yeesian/ArchGDAL.jl/pull/238#discussion_r707040320: they are mutually exclusive and I have not been able to construct any cases of fields that are null and unset, thank you!

Reply:
You’re welcome :-)

Reply:
Hey mathieu, the latest implementation in this PR (see https://github.com/yeesian/ArchGDAL.jl/pull/238/commits/958496206e31085aba53cb3ce5cb83aa09160a73) is based on https://github.com/yeesian/ArchGDAL.jl/pull/238#discussion_r708786099 and (i) assumes that the default value is only for writing fields and not when reading fields, (ii) returns `missing` rather than `nothing` for unset values. So I think it's the following table:

```
+-----------------+
| Read            | 
|-----------------|-----------+
| set and notnull |   value   |
|-----------------+-----------+
| everything else | `missing` |
|-----------------+-----------+
```

If you might feel strongly about returning `nothing` (rather than `missing`) for the `unset` case, I'm happy to revisit the implementation. Otherwise, my current thinking is to reserve `nothing` for the impossible case where we read a field that is both `unset` and `null`.

Reply:
> If you might feel strongly about returning nothing (rather than missing) for the unset case, I'm happy to revisit the implementation. Otherwise, my current thinking is to reserve nothing for the impossible case where we read a field that is both unset and null.

I think that:
- reading a field that is both **UNSET** and **NULL**, is not possible in GDAL's current implementation since those states correspond to different values of the same `Set struct` variable of an `OGRField` instance 
- we would loose the symmetry between reading and writing as, in my PR #243, I was about to map `nothing` to **NULL** and `missing` to **UNSET**, to give the ability to set either of those two states from a table source

But, if your are not convinced, no issue, I will map both `nothing` and `missing` to `UNSET`

Reply:
Thanks for re-kindling the debate @mathieu17g, and sorry it took me this long to think through it deeper: I agree with your reasoning and have reverted the code to the behavior described in https://github.com/yeesian/ArchGDAL.jl/pull/238#discussion_r699802843, namely,

```
+-------+---------+---------+
| Read |  null   | notnull |
+-------+---------+---------+
| set   | missing | value   |
+-------+---------+---------+
| unset | nothing | nothing |
+-------+---------+---------+

+---------+-------------+--------------+
|  Write  |  nullable   | notnullable  |
+---------+-------------+--------------+
| nothing |    unset    |     unset    |
+---------+-------------+--------------+
| missing |     null    | getdefault() |
+---------+-------------+--------------+
|  value  |    value    |    value     |
+---------+-------------+--------------+
```

where `getdefault()` returns `nothing` if there is no default value (which would cause it to be "unset" for notnullable columns).

---

That way, we
1. have a sufficiently rich set of Julian values (using both `missing` and `nothing`) to be able to reconstruct the state of a FeatureLayer (with the possible states of "unset" and "null").
2. support both (i) users who want to "fail early": use notnullable fields and error on `nothing`, and (ii) users who want to continue data processing with missing values: use nullable fields and propagate `missing` values.

Reply:
@mathieu17g @evetion I have updated the PR with https://github.com/yeesian/ArchGDAL.jl/pull/238/commits/08ffc271a64de12f352d12cb0eb4de3426ef1518 and documented it in https://yeesian.com/ArchGDAL.jl/previews/PR238/considerations/#Missing-and-Null-Semantics-in-GDAL -- thank you both so much! Can you have a look and let me know if this thread is good to be resolved?

Reply:
Resolving this thread since it's run its course. Feel free to open new comments / issues if there's anything unresolved.